### PR TITLE
[MM-33077] Add migrations for OAuthApps schema

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary
implements: https://mattermost.atlassian.net/browse/MM-33077

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33077

#### Release Note

```release-note
* Moves OAuthApps gorp driven migrations to schema-based migrations
```

### CODE REVIEW

File: db/migrations/mysql/000026_create_oauth_apps.up.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  CreatorId varchar(26) DEFAULT NULL,
+  CreateAt bigint(20) DEFAULT NULL,
+  UpdateAt bigint(20) DEFAULT NULL,
+  ClientSecret varchar(128) DEFAULT NULL,
+  Name varchar(64) DEFAULT NULL,
+  Description text,
+  CallbackUrls text,
+  Homepage text,
+  PRIMARY KEY (Id),
+  KEY idx_oauthapps_creator_id (CreatorId)
```
Comment (nronas) [inline db/migrations/mysql/000026_create_oauth_apps.up.sql:12]: We should make this a `create if not exists`. This will be skipped if the table exists but the index doesn't. Something like:

```sql
SET @preparedStatement = (SELECT IF(
    (
        SELECT COUNT(*) FROM INFORMATION_SCHEMA.statistics
        WHERE table_name = 'OAuthApps'
        AND table_schema = DATABASE()
        AND index_name = 'idx_oauthapps_creator_id'
    ) > 0,
    'SELECT 1',
    'CREATE INDEX ON OAuthApps(CreatorId);'
));

PREPARE createIndexIfNotExists FROM @preparedStatement;
EXECUTE createIndexIfNotExists;
DEALLOCATE PREPARE createIndexIfNotExists;
```

File: db/migrations/postgres/000046_create_oauth_apps.up.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    name VARCHAR(64),
+    description VARCHAR(512),
+    callbackurls VARCHAR(1024),
+    homepage VARCHAR(256)
+);
+
+CREATE INDEX IF NOT EXISTS idx_oauthapps_creator_id ON oauthapps (creatorid);
+
+ALTER TABLE oauthapps ADD COLUMN IF NOT EXISTS istrusted boolean DEFAULT false;
+ALTER TABLE oauthapps ADD COLUMN IF NOT EXISTS iconurl VARCHAR(512) DEFAULT '';
```
Comment (streamer45) [inline db/migrations/postgres/000046_create_oauth_apps.up.sql:16]: In the current system we don't set defaults on creation and usually we prefer to avoid them. I am wondering if we should change these as well.

File: db/migrations/postgres/000046_create_oauth_apps.up.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    id VARCHAR(26) PRIMARY KEY,
+    creatorid VARCHAR(26),
+    createat bigint,
+    updateat bigint,
+    clientsecret VARCHAR(128),
+    name VARCHAR(64),
+    description VARCHAR(512),
+    callbackurls VARCHAR(1024),
+    homepage VARCHAR(256)
+);
```
Comment (streamer45) [inline db/migrations/postgres/000046_create_oauth_apps.up.sql:11]: Another general question: I am wondering if it would be simpler for the table creation code to have the most recent version of the table. I am thinking it would be the same logic wise.

Comment (isacikgoz) [reply db/migrations/postgres/000046_create_oauth_apps.up.sql:11]: But in that case we would lose the history right? And if we are going to include the down scripts with the reverse steps it wouldn't be consistent. cc @mgdelacroix

File: db/migrations/postgres/000046_create_oauth_apps.up.sql (hunk 2 above)
Comment (isacikgoz) [reply db/migrations/postgres/000046_create_oauth_apps.up.sql:16]: Hmm, makes sense 👍

**Review by isacikgoz — APPROVED** [review]
LGTM 😛

File: store/sqlstore/upgrade.go [hunk 4]
```diff
@@ -303,11 +303,6 @@ func upgradeDatabaseToVersion33(sqlStore *SqlStore) {
 			}
 		}
 
-		sqlStore.CreateColumnIfNotExists("OAuthApps", "IsTrusted", "tinyint(1)", "boolean", "0")
-		sqlStore.CreateColumnIfNotExists("OAuthApps", "IconURL", "varchar(512)", "varchar(512)", "")
-
-		sqlStore.CreateColumnIfNotExists("OutgoingWebhooks", "TriggerWhen", "tinyint", "integer", "0")
```
Comment (streamer45) [inline store/sqlstore/upgrade.go:309]: How is this related?

File: store/sqlstore/oauth_store.go [hunk 5]
```diff
@@ -53,12 +53,6 @@ func newSqlOAuthStore(sqlStore *SqlStore) store.OAuthStore {
 	return as
 }
 
-func (as SqlOAuthStore) createIndexesIfNotExists() {
-	as.CreateIndexIfNotExists("idx_oauthapps_creator_id", "OAuthApps", "CreatorId")
-	as.CreateIndexIfNotExists("idx_oauthaccessdata_user_id", "OAuthAccessData", "UserId")
-	as.CreateIndexIfNotExists("idx_oauthaccessdata_refresh_token", "OAuthAccessData", "RefreshToken")
```
Comment (streamer45) [inline store/sqlstore/oauth_store.go:59]: We are removing stuff for `OAuthAccessData` but the PR merely focuses on `OAuthApps`.

File: store/sqlstore/store.go [hunk 6]
```diff
@@ -259,7 +259,6 @@ func New(settings model.SqlSettings, metrics einterfaces.MetricsInterface) *SqlS
 
 	store.stores.channel.(*SqlChannelStore).createIndexesIfNotExists()
 	store.stores.retentionPolicy.(*SqlRetentionPolicyStore).createIndexesIfNotExists()
-	store.stores.oauth.(*SqlOAuthStore).createIndexesIfNotExists()
```
Comment (streamer45) [inline store/sqlstore/store.go:262]: This also affects other tables that are not covered in here.

File: store/sqlstore/oauth_store.go (hunk 5 above)
Comment (isacikgoz) [reply store/sqlstore/oauth_store.go:59]: These are forgotten
- https://github.com/mattermost/mattermost-server/blob/6ac9cdc98767dc7adaa7dc8b2792c9d2c77f5c0f/db/migrations/mysql/000025_create_oauth_access_data.up.sql#L49
- https://github.com/mattermost/mattermost-server/blob/6ac9cdc98767dc7adaa7dc8b2792c9d2c77f5c0f/db/migrations/mysql/000025_create_oauth_access_data.up.sql#L139

File: store/sqlstore/store.go (hunk 6 above)
Comment (isacikgoz) [reply store/sqlstore/store.go:262]: Related with above.

File: store/sqlstore/upgrade.go (hunk 4 above)
Comment (isacikgoz) [reply store/sqlstore/upgrade.go:309]: I literally have no idea, It seems like added before as well:
https://github.com/mattermost/mattermost-server/blob/4739e93c16a0d0d884143d5f12c5ee710f6c08cb/db/migrations/mysql/000014_create_outgoing_webhooks.up.sql#L39

**Review by streamer45 — APPROVED** [review]
Discussed internally. Approving :+1:
